File: android/app/src/main/java/com/dyjung/logindemo/presentation/auth/RegisterViewModel.kt
package com.dyjung.logindemo.presentation.auth

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.dyjung.logindemo.core.validation.EmailValidator
import com.dyjung.logindemo.core.validation.NameValidator
import com.dyjung.logindemo.core.validation.PasswordValidator
import com.dyjung.logindemo.domain.usecase.CheckEmailUseCase
import com.dyjung.logindemo.domain.usecase.RegisterUseCase
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import javax.inject.Inject

/**
 * 회원가입 화면 ViewModel
 * iOS의 RegisterViewModel에 대응
 */
@HiltViewModel
class RegisterViewModel @Inject constructor(
    private val registerUseCase: RegisterUseCase,
    private val checkEmailUseCase: CheckEmailUseCase
) : ViewModel() {

    private val _uiState = MutableStateFlow(RegisterUiState())
    val uiState: StateFlow<RegisterUiState> = _uiState.asStateFlow()

    private var emailCheckJob: Job? = null

    fun onEmailChange(email: String) {
        val validation = EmailValidator.validate(email)
        _uiState.update { state ->
            state.copy(
                email = email,
                emailError = validation.errorMessage,
                isEmailAvailable = null
            )
        }

        // 이메일 형식이 올바른 경우 중복 확인
        if (validation.isValid && email.isNotBlank()) {
            checkEmailAvailability(email)
        }
    }

    private fun checkEmailAvailability(email: String) {
        emailCheckJob?.cancel()

        emailCheckJob = viewModelScope.launch {
            // 0.5초 디바운스
            delay(500)

            _uiState.update { it.copy(isCheckingEmail = true) }

            try {
                val available = checkEmailUseCase(email)
                _uiState.update { state ->
                    state.copy(
                        isEmailAvailable = available,
                        emailError = if (!available) "이미 사용 중인 이메일입니다" else null,
                        isCheckingEmail = false
                    )
                }
            } catch (e: Exception) {
                _uiState.update { it.copy(isCheckingEmail = false) }
            }
        }
    }

    fun onPasswordChange(password: String) {
        val validation = PasswordValidator.validate(password)
        _uiState.update { state ->
            state.copy(
                password = password,
                passwordError = validation.errorMessage
            )
        }

        // 비밀번호 확인 필드도 재검증
        if (_uiState.value.confirmPassword.isNotBlank()) {
            validateConfirmPassword()
        }
    }

    fun onConfirmPasswordChange(confirmPassword: String) {
        _uiState.update { it.copy(confirmPassword = confirmPassword) }
        validateConfirmPassword()
    }

    private fun validateConfirmPassword() {
        val validation = PasswordValidator.validateConfirmation(
            _uiState.value.password,
            _uiState.value.confirmPassword
        )
        _uiState.update { it.copy(confirmPasswordError = validation.errorMessage) }
    }

    fun onNameChange(name: String) {
        val validation = NameValidator.validate(name)
        _uiState.update { state ->
            state.copy(
                name = name,
                nameError = validation.errorMessage
            )
        }
    }

    fun onTermsChange(agreed: Boolean) {
        _uiState.update { it.copy(agreedToTerms = agreed) }
    }

    fun onPrivacyChange(agreed: Boolean) {
        _uiState.update { it.copy(agreedToPrivacy = agreed) }
    }

    fun toggleAllAgreements() {
        val newValue = !(_uiState.value.agreedToTerms && _uiState.value.agreedToPrivacy)
        _uiState.update {
            it.copy(
                agreedToTerms = newValue,
                agreedToPrivacy = newValue
            )
        }
    }

    fun register(onSuccess: () -> Unit) {
        if (!_uiState.value.isRegisterEnabled) return

        viewModelScope.launch {
            _uiState.update { it.copy(isLoading = true, error = null) }

            try {
                registerUseCase(
                    email = _uiState.value.email,
                    password = _uiState.value.password,
                    name = _uiState.value.name,
                    agreedToTerms = _uiState.value.agreedToTerms,
                    agreedToPrivacy = _uiState.value.agreedToPrivacy
                )
                onSuccess()
            } catch (e: Exception) {
                _uiState.update { it.copy(error = e.message ?: "회원가입에 실패했습니다") }
            } finally {
                _uiState.update { it.copy(isLoading = false) }
            }
        }
    }

    fun clearError() {
        _uiState.update { it.copy(error = null) }
    }

    override fun onCleared() {
        super.onCleared()
        emailCheckJob?.cancel()
    }
}

/**
 * 회원가입 화면 UI 상태
 */
data class RegisterUiState(
    val email: String = "",
    val password: String = "",
    val confirmPassword: String = "",
    val name: String = "",
    val agreedToTerms: Boolean = false,
    val agreedToPrivacy: Boolean = false,
    val emailError: String? = null,
    val isEmailAvailable: Boolean? = null,
    val passwordError: String? = null,
    val confirmPasswordError: String? = null,
    val nameError: String? = null,
    val isLoading: Boolean = false,
    val isCheckingEmail: Boolean = false,
    val error: String? = null
) {
    val isRegisterEnabled: Boolean
        get() = email.isNotBlank() &&
                emailError == null &&
                isEmailAvailable == true &&
                password.isNotBlank() &&
                passwordError == null &&
                confirmPassword.isNotBlank() &&
                confirmPasswordError == null &&
                name.isNotBlank() &&
                nameError == null &&
                agreedToTerms &&
                agreedToPrivacy &&
                !isLoading
}

File: ios/LoginDemo/Presentation/Auth/Register/RegisterViewModel.swift
//
//  RegisterViewModel.swift
//  LoginDemo
//
//  회원가입 화면 ViewModel
//

import Foundation
import Observation

/// 회원가입 화면 ViewModel
@Observable
@MainActor
final class RegisterViewModel {
    // MARK: - Properties

    /// 이메일 입력값
    var email: String = "" {
        didSet { handleEmailChange() }
    }

    /// 비밀번호 입력값
    var password: String = "" {
        didSet { validatePassword() }
    }

    /// 비밀번호 확인 입력값
    var confirmPassword: String = "" {
        didSet { validateConfirmPassword() }
    }

    /// 닉네임 입력값 (기존 name)
    var nickname: String = "" {
        didSet { validateNickname() }
    }

    /// 이용약관 동의 여부
    var agreedToTerms: Bool = false

    /// 개인정보처리방침 동의 여부
    var agreedToPrivacy: Bool = false

    /// 마케팅 수신 동의 여부
    var marketingConsent: Bool = false

    /// 이메일 오류 메시지
    private(set) var emailError: String?

    /// 비밀번호 오류 메시지
    private(set) var passwordError: String?

    /// 비밀번호 확인 오류 메시지
    private(set) var confirmPasswordError: String?

    /// 닉네임 오류 메시지 (기존 nameError)
    private(set) var nicknameError: String?

    /// 로딩 상태
    private(set) var isLoading: Bool = false

    /// 에러
    var error: Error?

    /// 회원가입 버튼 활성화 여부
    var isRegisterEnabled: Bool {
        !email.isEmpty &&
        emailError == nil &&
        !password.isEmpty &&
        passwordError == nil &&
        !confirmPassword.isEmpty &&
        confirmPasswordError == nil &&
        !nickname.isEmpty &&
        nicknameError == nil &&
        agreedToTerms &&
        agreedToPrivacy &&
        !isLoading
    }

    private let registerUseCase: RegisterUseCase

    // MARK: - Initialization

    init(registerUseCase: RegisterUseCase) {
        self.registerUseCase = registerUseCase
    }

    // MARK: - Resource Cleanup

    /// 리소스 정리 (화면 전환 시 호출)
    func cleanup() {
        isLoading = false
    }

    // MARK: - Methods

    /// 회원가입 실행
    /// - Parameter appState: 앱 상태
    func register(appState: AppState) async {
        guard isRegisterEnabled else { return }

        isLoading = true
        error = nil

        do {
            let user = try await registerUseCase.execute(
                email: email,
                password: password,
                nickname: nickname,
                marketingConsent: marketingConsent
            )

            appState.handleLoginSuccess(user: user)
        } catch {
            self.error = error
        }

        isLoading = false
    }

    /// 전체 동의 토글
    func toggleAllAgreements() {
        let newValue = !(agreedToTerms && agreedToPrivacy && marketingConsent)
        agreedToTerms = newValue
        agreedToPrivacy = newValue
        marketingConsent = newValue
    }

    // MARK: - Validation

    private func handleEmailChange() {
        validateEmail()
    }

    private func validateEmail() {
        let result = EmailValidator.validate(email)
        emailError = result.errorMessage
    }

    private func validatePassword() {
        let result = PasswordValidator.validate(password)
        passwordError = result.errorMessage

        // 비밀번호 확인 필드도 재검증
        if !confirmPassword.isEmpty {
            validateConfirmPassword()
        }
    }

    private func validateConfirmPassword() {
        let result = PasswordValidator.validateConfirmation(
            password: password,
            confirmation: confirmPassword
        )
        confirmPasswordError = result.errorMessage
    }

    private func validateNickname() {
        if nickname.count < 2 {
            nicknameError = "닉네임은 2자 이상이어야 합니다"
        } else {
            nicknameError = nil
        }
    }

    /// 입력값 초기화
    func reset() {
        email = ""
        password = ""
        confirmPassword = ""
        nickname = ""
        agreedToTerms = false
        agreedToPrivacy = false
        marketingConsent = false
        emailError = nil
        passwordError = nil
        confirmPasswordError = nil
        nicknameError = nil
        error = nil
    }
}
